File: day24/convertBiNode.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# 递归
# class Solution:
#     def convertBiNode(self, root: TreeNode) -> TreeNode:
#         res = TreeNode(0)
#         cur = res
#         def inOrder(cur, root):
#             if root == None: return cur
#             cur = inOrder(cur, root.left)
#             root.left = None
#             cur.right = root
#             cur = cur.right
#             cur = inOrder(cur, root.right)
#             return cur
#         cur = inOrder(cur, root)
#         return res.right

# 循环
class Solution:
    def convertBiNode(self, root: TreeNode) -> TreeNode:
        res = TreeNode(0)
        cur = res
        stack = collections.deque()
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                root = stack.pop()
                root.left = None
                cur.right = root
                cur = cur.right
                root = root.right
        return res.right

File: day24/convertBiNode.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 循环
class Solution {
    
    public TreeNode convertBiNode(TreeNode root) {
        TreeNode res = new TreeNode(0);
        TreeNode cur = res;
        Stack<TreeNode> stack = new Stack<>();
        while (root != null || !stack.empty()) {
            if (root != null) {
                stack.push(root);
                root = root.left;
            } else {
                root = stack.pop();
                root.left = null;
                cur.right = root;
                cur = cur.right;
                root = root.right;
            }
        }
        return res.right;
    }

}

// 递归
// class Solution {
    
//     public TreeNode convertBiNode(TreeNode root) {
//         TreeNode res = new TreeNode(0);
//         inOrder(root, res);
//         return res.right;
//     }

//     public TreeNode inOrder(TreeNode root, TreeNode cur) {
//         if (root == null) return cur;
//         cur = inOrder(root.left, cur);
//         root.left = null;
//         cur.right = root;
//         cur = cur.right;
//         cur = inOrder(root.right, cur);
//         return cur;
//     }

// }
